sharding/nodesCoordinator/indexHashedNodesCoordinator.go

```diff
@@ -829,11 +832,33 @@ func (ihnc *indexHashedNodesCoordinator) computeNodesConfigFromList(
 }
 
 func (ihnc *indexHashedNodesCoordinator) addValidatorToPreviousMap(
+	previousEpochConfig *epochNodesConfig,
 	eligibleMap map[uint32][]Validator,
+	waitingMap map[uint32][]Validator,
 	currentValidator *validator,
 	currentValidatorShardId uint32,
 ) {
-	eligibleMap[currentValidatorShardId] = append(eligibleMap[currentValidatorShardId], currentValidator)
+	if !ihnc.flagStakingV4Started.IsSet() {
+		eligibleMap[currentValidatorShardId] = append(eligibleMap[currentValidatorShardId], currentValidator)
+		return
+	}
+
+	found, shardId := searchInMap(previousEpochConfig.eligibleMap, currentValidator.PubKey())
+	if found {
+		log.Debug("leaving node found in", "list", "eligible", "shardId", shardId)
+		eligibleMap[shardId] = append(eligibleMap[currentValidatorShardId], currentValidator)
+		return
+	}
+
+	found, shardId = searchInMap(previousEpochConfig.waitingMap, currentValidator.PubKey())
+	if found {
+		log.Debug("leaving node found in", "list", "waiting", "shardId", shardId)
+		waitingMap[shardId] = append(waitingMap[currentValidatorShardId], currentValidator)
+		return
+	}
+
+	log.Debug("leaving node not in eligible or waiting, probably was in auction/inactive/jailed",
```

Comment:
currentValidatorShardId instead shardId. I think in this point shardId will have not value set, probably 0, as found were false in both searching above.